engine/consensus/dkg/processor/engine.go

```diff
@@ -0,0 +1,195 @@
+// DKG protocol should use the same _ordered_ list of committee identifiers,
+// because senders and recipients are identified by their index in this list.
+// The msgCh is used to forward incoming DKG messages to the consumer, such as
+// the Controller implemented in the DKG module.
+func New(
+	logger zerolog.Logger,
+	net module.Network,
+	me module.Local,
+	msgCh chan dkg.DKGMessage,
+	committee flow.IdentifierList,
+	epochID flow.Identifier) (*Engine, error) {
+
+	log := logger.With().Str("engine", "dkg-processor").Logger()
+
+	index := -1
+	for i, id := range committee {
+		if id == me.NodeID() {
+			index = i
+		}
+	}
+	if index < 0 {
+		return nil, fmt.Errorf("dkg-processor engine id does not belong to dkg committee")
+	}
+
+	eng := Engine{
+		unit:      engine.NewUnit(),
+		log:       log,
+		me:        me,
+		msgCh:     msgCh,
+		committee: committee,
+		myIndex:   index,
+		epochID:   epochID,
+	}
+
+	var err error
+	eng.conduit, err = net.Register(engine.DKGCommittee, &eng)
+	if err != nil {
+		return nil, fmt.Errorf("could not register dkg-processor engine: %w", err)
+	}
+
+	return &eng, nil
+}
+
+// SetEpoch changes the epoch ID of the engine.
+func (e *Engine) SetEpoch(epochID flow.Identifier) {
+	e.epochIDLock.Lock()
+	defer e.epochIDLock.Unlock()
+	e.epochID = epochID
+}
+
+// Ready implements the module ReadyDoneAware interface. It returns a channel
+// that will close when the engine has successfully
+// started.
+func (e *Engine) Ready() <-chan struct{} {
+	return e.unit.Ready()
+}
+
+// Done implements the module ReadyDoneAware interface. It returns a channel
+// that will close when the engine has successfully stopped.
+func (e *Engine) Done() <-chan struct{} {
+	return e.unit.Done()
+}
+
+// SubmitLocal implements the network Engine interface
+func (e *Engine) SubmitLocal(event interface{}) {
+	e.Submit(e.me.NodeID(), event)
+}
+
+// Submit implements the network Engine interface
+func (e *Engine) Submit(originID flow.Identifier, event interface{}) {
+	e.unit.Launch(func() {
+		err := e.Process(originID, event)
+		if err != nil {
+			engine.LogError(e.log, err)
+		}
+	})
+}
+
+// ProcessLocal implements the network Engine interface
+func (e *Engine) ProcessLocal(event interface{}) error {
+	return e.Process(e.me.NodeID(), event)
+}
+
+// Process implements the network Engine interface
+func (e *Engine) Process(originID flow.Identifier, event interface{}) error {
+	return e.unit.Do(func() error {
+		return e.process(originID, event)
+	})
+}
+
+func (e *Engine) process(originID flow.Identifier, event interface{}) error {
+	switch v := event.(type) {
+	case dkg.DKGMessage:
+		return e.onDKGMessage(originID, v)
+	default:
+		return fmt.Errorf("invalid event type (%T)", event)
+	}
+}
+
+func (e *Engine) onDKGMessage(originID flow.Identifier, msg dkg.DKGMessage) error {
+
+	// check that the message corresponds to the current epoch
+	e.epochIDLock.Lock()
+	defer e.epochIDLock.Unlock()
```

Comment:
The `engine.Unit` has an embedded mutex that can be used for synchronizing the engine -- we can use that here. 
```suggestion
	e.unit.Lock()
	defer e.unit.Unlock()
```